sdk/nodejs/policy/proxy.ts

```diff
@@ -0,0 +1,106 @@
+// Copyright 2016-2019, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+export function proxy<T>(toProxy: any): any {
+    return proxyHelper(toProxy, []);
+}
+
+function proxyHelper<T>(toProxy: any, props: (keyof T)[]): any {
+    if (!(toProxy instanceof Object)) {
+        return toProxy;
+    }
+
+    return new Proxy(toProxy, {
+        get: (obj, prop: keyof T) => {
+            const newProps = props.concat([prop]);
+            const field = obj[prop];
+            if (isUnknown(field)) {
+                throw new UnknownValueError<T>(field, newProps);
+            }
+            return proxyHelper(field, newProps);
+        },
+    });
+}
+
+export class UnknownValueError<T> extends Error {
+    constructor(public readonly unknownType: string, public readonly props: (keyof T)[]) {
```

Comment:
query: does anyone uses the args outside of this error?  If not, consider not making them properties.

Reply:
ah, i see you use them for testing purposes.

Reply:
I don't think I know what this means, but shortening the call to super means that we have to have local variables before `super`, which means that we need to set these properties manually vs having them baked into the constructor.

Reply:
so, `constructor(public whatever)` means you're declaring both a parameter and a property on hte class.  I was asking if you actually needed the property or not.  I see you actually use the property later on in tests, so this is ok.